test/preview/components/generators/test_hugging_face_tgi.py

```diff
@@ -0,0 +1,295 @@
+
+from haystack.preview.components.generators import HuggingFaceTGIGenerator
+from haystack.preview.dataclasses import StreamingChunk
+
+
+@pytest.fixture
+def mock_check_valid_model():
+    with patch(
+        "haystack.preview.components.generators.hugging_face_tgi.check_valid_model", MagicMock(return_value=None)
+    ) as mock:
+        yield mock
+
+
+@pytest.fixture
+def mock_text_generation():
+    with patch("huggingface_hub.InferenceClient.text_generation", autospec=True) as mock_from_pretrained:
+        mock_response = Mock()
+        mock_response.generated_text = "I'm fine, thanks."
+        details = Mock()
+        details.finish_reason = MagicMock(field1="value")
+        details.tokens = [1, 2, 3]
+        mock_response.details = details
+        mock_from_pretrained.return_value = mock_response
+        yield mock_from_pretrained
+
+
+# used to test serialization of streaming_callback
+def streaming_callback_handler(x):
+    return x
+
+
+class TestHuggingFaceTGIGenerator:
+    @pytest.mark.unit
+    def test_initialize_with_valid_model_and_generation_parameters(self, mock_check_valid_model, mock_auto_tokenizer):
+        model = "HuggingFaceH4/zephyr-7b-alpha"
+        generation_kwargs = {"n": 1}
+        stop_words = ["stop"]
+        streaming_callback = None
+
+        generator = HuggingFaceTGIGenerator(
+            model=model,
+            url=None,
+            token=None,
+            generation_kwargs=generation_kwargs,
+            stop_words=stop_words,
+            streaming_callback=streaming_callback,
+        )
+
+        assert generator.model == model
+        assert generator.generation_kwargs == {**generation_kwargs, **{"stop_sequences": ["stop"]}}
+        assert generator.tokenizer is None
+        assert generator.client is not None
+        assert generator.streaming_callback == streaming_callback
+
+    @pytest.mark.unit
+    def test_to_dict(self, mock_check_valid_model, mock_auto_tokenizer):
+        # Initialize the HuggingFaceRemoteGenerator object with valid parameters
+        generator = HuggingFaceTGIGenerator(
+            token="token", generation_kwargs={"n": 5}, stop_words=["stop", "words"], streaming_callback=lambda x: x
+        )
+
+        # Call the to_dict method
+        result = generator.to_dict()
+        init_params = result["init_parameters"]
+
+        # Assert that the init_params dictionary contains the expected keys and values
+        assert init_params["model"] == "mistralai/Mistral-7B-v0.1"
+        assert not init_params["token"]
+        assert init_params["generation_kwargs"] == {"n": 5, "stop_sequences": ["stop", "words"]}
+
+    @pytest.mark.unit
+    def test_from_dict(self, mock_check_valid_model, mock_auto_tokenizer):
+        generator = HuggingFaceTGIGenerator(
+            model="mistralai/Mistral-7B-v0.1",
+            generation_kwargs={"n": 5},
+            stop_words=["stop", "words"],
+            streaming_callback=streaming_callback_handler,
```

Comment:
Using `streaming_callback=lambda x: x` would make the test self-contained.

In any case, we should choose whether to use a lambda everywhere or to use the `streaming_callback_handler` function everywhere.

Reply:
I'd rather use `streaming_callback_handler` because we then check if it was properly deserialized (loc 91) Cleaner I think